Add copy to clipboard btn in creation success page

skbro opened the pull request:
Closes #132 
Copy to Clipboard Button added using clipboard API.
![Screenshot from 2019-10-24 00-01-40](https://user-images.githubusercontent.com/33025439/67424138-39f92c80-f5f3-11e9-910a-761633953a1c.png)

ThiefMaster commented:
Thanks! I think it would be nicer if the button was either attached to the field or inside it.

What about using a readonly Input with an attached button? Check https://react.semantic-ui.com/elements/input/ for some examples.

Just please don't go for the huge copy button in one of their examples (without the text it might be nice though).

skbro commented:
Okay, sure.

skbro commented:
Closes Issue #132 
Changed Grid Component to Disabled Input 
![Screenshot from 2019-10-24 00-49-21](https://user-images.githubusercontent.com/33025439/67431302-9ebb8380-f601-11e9-80c1-dba61d97d574.png)

ThiefMaster commented on newdle/client/src/components/creation/CreationSuccessPage.js:
```diff
@@ -38,7 +38,17 @@ export default function CreationSuccessPage() {
-        <div className={styles['newdle-link']}>{newdle.url}</div>
+        <Input
+          className={styles['newdle-link']}
+          disabled
```
I think `readOnly` would be more appropriate here, but I'll have to try it out and see it in action. Could even select the contents of the field when focusing it..

skbro replied:
Yeah, the disabled field is having some problems to select the contents. Let me try with readOnly and I will push accordingly.

ThiefMaster commented on newdle/client/src/components/creation/CreationSuccessPage.js:
```diff
@@ -38,7 +38,17 @@ export default function CreationSuccessPage() {
+          action={{
+            icon: 'copy',
+            onClick: async () => {
+              await navigator.clipboard.writeText(newdle.url);
```
You don't need to `await` here (and thus no `async` needed in the line above) since there's nothing you do after copying the link anyway.

Could be nice to show some kind of "copied" message though - but not sure if it's really needed (feel free to give it a try though), or use a CSS animation to style the button a bit differently to make it clear that something happened when you clicked it.

skbro commented:
Added a little CSS animation to show something happened when the copy button was clicked. I tried adding the popup component to show a nice copied message, but I think state is required for that.

ThiefMaster commented on newdle/client/src/components/creation/CreationSuccessPage.js:
```diff
@@ -38,7 +38,19 @@ export default function CreationSuccessPage() {
+            title="Copy to clipboard"
+            onClick={() => navigator.clipboard.writeText(newdle.url)}
+            icon="copy"
+          ></Button>
```
`<Button ... />`

ThiefMaster commented on newdle/client/src/components/creation/CreationSuccessPage.js:
```diff
@@ -38,7 +38,19 @@ export default function CreationSuccessPage() {
+          value={newdle.url}
+          action={{icon: 'copy'}}
+        >
+          <input />
```
that looks wrong, do we need that?!

skbro replied:
Yeah, I think that is not required. Going to remove it

ThiefMaster commented on newdle/client/src/components/creation/CreationSuccessPage.js:
```diff
@@ -38,7 +38,19 @@ export default function CreationSuccessPage() {
-        <div className={styles['newdle-link']}>{newdle.url}</div>
+        <Input
+          className={styles['newdle-link']}
+          readOnly={true}
```
just `readOnly`

skbro replied:
Sure

ThiefMaster commented on newdle/client/src/components/creation/CreationSuccessPage.js:
```diff
@@ -38,7 +38,19 @@ export default function CreationSuccessPage() {
+          <input />
+          <Button
+            title="Copy to clipboard"
+            onClick={() => navigator.clipboard.writeText(newdle.url)}
```
awaiting is only needed when you actually do something after it finished

ThiefMaster commented on newdle/client/src/components/creation/CreationSuccessPage.module.scss:
```diff
@@ -10,9 +10,17 @@
+    color: black !important;
+  }
+
+  button:active {
```
not a big fan of this "animation" style... maybe using state to show a Popup is nicer after all? maybe we can actually go back to only having the copy button in this PR and then, if you want, you send another one that adds the popup?

skbro replied:
Let me try adding popup again and remove the animation.

skbro commented:
![Screenshot from 2019-10-25 09-59-32](https://user-images.githubusercontent.com/33025439/67543424-ef6cd280-f70e-11e9-8af9-7f8b08715d6e.png)
Added Popup, removed all custom animations and other reviewed problems.

pferreir commented on newdle/client/src/components/creation/CreationSuccessPage.js:
```diff
@@ -38,7 +38,26 @@ export default function CreationSuccessPage() {
+              content="Copied!"
+              on="click"
+              position="top center"
+              inverted
```
I understand that it doesn't make much difference because the copying itself is an instantaneous operation. However, the "right" way to do this would be to only display the `<Popup />` once `writeText(...)` succeeds. I'm not sure under which conditions `writeText(...)` might fail (the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/writeText) talk of not having permission to access the clipboard), but maybe it would be worth making the `<Popup />` controlled and adding an additional state to take care of that.
@ThiefMaster opinions?

ThiefMaster replied:
Agreed, for showing the popup I think using state that's only set after copying would be nicer.. (and then maybe using setTimeout to hide it after a second or two).

We should probably also try..catch around the call to avoid an error when the promise does get rejected. Probably no need to actually do anything in there for now, since it seems unlikely that it happens..

ThiefMaster replied:
I tested in firefox with `dom.events.asyncClipboard` disabled and the button gets hidden as expected. I think for now it's fine, if we find out that there are cases where `navigator.clipboard` is defined but copying still fails we can still add error handling there

pferreir commented on newdle/client/src/components/creation/CreationSuccessPage.js:
```diff
@@ -38,7 +38,26 @@ export default function CreationSuccessPage() {
+              position="top center"
+              inverted
+              trigger={
+                <Button
```
The asynchronous Clipboard API is not supported on Edge, so I would probably hide this button in that case. We can just check the truthiness of `navigator.clipboard`, I believe. Anyway, it would be nice if we test it with Edge before we merge.

ThiefMaster replied:
That's weird, generally localhost should be exempt from "https-only" restrictions. If not I'd consider it a bug in the browser...

skbro replied:
[Found on stackoverflow](https://stackoverflow.com/a/51823007)

ThiefMaster replied:
oops, I guess I misread your last comment as "doesn't work on localhost".

ThiefMaster replied:
`action={navigator.clipboard && <...>}` should be fine, we did the same in the indico PR linked in a commend of the original issue

ThiefMaster commented:
Thanks a lot, it looks good now. I added some small improvements and will now merge it.

Just a tip for the future: Create a branch for your PRs instaed of working directly on your master branch. That way you keep master clean.

skbro commented:
Yeah, I will make sure of that.
Thank you for the awesome learning experience.

pferreir commented:
Thanks a lot for your help, @skbro ! It's really appreciated!
